src/fnargs.c

```diff
@@ -221,6 +250,31 @@ int prepare_fn_args_specs(int func_id, const struct mass_attacher_func_info *fin
 			true_len = (is_vararg || btf_is_ptr(t)) ? 8 : t->size;
 			data_len = min(true_len, env.args_max_sized_arg_size);
 
+#ifdef __aarch64__
+			/* Special handling for structs in ARM64 */
+			if (btf_is_struct(t) && t->size <= 16) {
+				if (t->size <= 8 && is_arg_in_reg(reg_idx, &reg1_name)) {
+					/* Fits in one register */
+					spec->arg_flags |= data_len;
+					spec->arg_flags |= FNARGS_REG << FNARGS_LOC_SHIFT;
+					spec->arg_flags |= reg_idx << FNARGS_REGIDX_SHIFT;
+					dlog("%s", reg1_name);
+					reg_idx += 1;
+				} else if (t->size <= 16 && 
+						is_arg_in_reg(reg_idx, &reg1_name) && 
+						is_arg_in_reg(reg_idx + 1, &reg2_name)) {
+					/* Passed in a pair of registers */
+					spec->arg_flags |= data_len;
+					spec->arg_flags |= FNARGS_REG_PAIR << FNARGS_LOC_SHIFT;
+					spec->arg_flags |= reg_idx << FNARGS_REGIDX_SHIFT;
+					reg_idx += 2;
+					dlog("%s:%s", reg1_name, reg2_name);
+				} else {
+					/* Fallback to stack */
+					goto use_stack;
+				}
+			} else
+#endif
```

Comment:
we have exactly the same logic for x86-64, no need for this duplication

Reply:
addressed in https://github.com/anakryiko/retsnoop/pull/89/commits/0aacc5864af055e4a36834d8c914427fb7f26295